### PR TITLE
Add form validation to add user

### PR DESCRIPTION
**Pull request by kklamberty**
This commit addresses #28 and includes several end-to-end tests of the form validation. There are a few tests missing (user name being too short or too long), but give a good idea about how this works. We should probably add those, but this is real progress toward this issue and could maybe close it.

### CODE REVIEW

**Comment by kklamberty** [comment]
E2E seem to pass for me as well as the tests run for a successful build, by the way.

File: client/src/app/users/name.validator.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { FormControl } from '@angular/forms';
+
+export class NameValidator {
```
Comment (NicMcPhee) [inline client/src/app/users/name.validator.ts:3]: Nice to have included an example of a custom validator. Your comments, though, imply that it's not "working as it should"? Does it not actually work? Are there issues we should try to address?

**Review by NicMcPhee — CHANGES_REQUESTED** [review]
This all looks really cool, and I'm actually happy if you just merge it as is. There are a couple of small refactorings, though, that would both help improve the readability of the code and demonstrate how they can/should clean up duplication in their code is well.

Thanks!

File: client/e2e/user-list.e2e-spec.ts [hunk 2]
```diff
@@ -2,21 +2,25 @@ import {UserPage} from './user-list.po';
 import {browser, protractor, element, by} from 'protractor';
 import {Key} from 'selenium-webdriver';
 
+// This line (combined with the function that follows) is here for us
```
[the remark this answers is not in the document]
Comment (kklamberty) [reply client/e2e/user-list.e2e-spec.ts:5]: Yes

File: client/e2e/user-list.e2e-spec.ts [hunk 3]
```diff
@@ -120,6 +124,84 @@ describe('User list', () => {
         expect(page.elementExistsWithCss('add-user')).toBeTruthy('There should be a modal window now');
     });
 
+    it('Should allow us to put information into the fields of the add user dialog', () => {
+        page.navigateTo();
+        page.click('addNewUser');
+        expect(page.field('nameField').isPresent()).toBeTruthy('There should be a name field');
+        page.field('nameField').sendKeys('Dana Jones');
+        expect(element(by.id('ageField')).isPresent()).toBeTruthy('There should be an age field');
+        // Need to use backspace because the default value is -1. If that changes, this will change too.
```
[the remark this answers is not in the document]
Comment (kklamberty) [reply client/e2e/user-list.e2e-spec.ts]: Maybe we can clear the field from our e2e test and still avoid putting the test functionality itself into the page objects.

File: client/e2e/user-list.e2e-spec.ts [hunk 4]
```diff
@@ -120,6 +124,84 @@ describe('User list', () => {
         expect(page.elementExistsWithCss('add-user')).toBeTruthy('There should be a modal window now');
     });
 
+    it('Should allow us to put information into the fields of the add user dialog', () => {
+        page.navigateTo();
+        page.click('addNewUser');
+        expect(page.field('nameField').isPresent()).toBeTruthy('There should be a name field');
```
[the remark this answers is not in the document]
Comment (kklamberty) [reply client/e2e/user-list.e2e-spec.ts]: Yes - let's put that function in the e2e test (not in page objects)

[the remark this answers is not in the document]
Comment (kklamberty) [reply client/e2e/user-list.e2e-spec.ts]: For now, let's not change this. Thinking more about this, a failure is easier to detect when we know from which test it failed.

File: client/src/app/custom.module.ts [hunk 5]
```diff
@@ -21,6 +21,8 @@ const FLEX_LAYOUT_MODULES: any[] = [
 const ANGULAR_MODULES: any[] = [
     BrowserAnimationsModule,
     FormsModule,
+    ReactiveFormsModule
+
```
[the remark this answers is not in the document]
Comment (kklamberty) [reply client/src/app/custom.module.ts]:
```suggestion
```

**Review by NicMcPhee — APPROVED** [review]
@kklamberty and I spent several hours on some refactoring and improvements that address my various comments here. I think we're good to commit!

**Comment by kklamberty** [comment]
Closes #28

**Pull request merged**